contrib/charts/cert-manager/values.yaml

```diff
@@ -6,29 +6,41 @@ replicaCount: 1
 image:
   repository: quay.io/jetstack/cert-manager-controller
   tag: v0.2.2
-  pullPolicy: Always
+  pullPolicy: IfNotPresent
 
 createCustomResource: true
 
 rbac:
-  enabled: true
+  create: true
+  serviceAccountName: default
```

Comment:
Not sure on best practice here, but are we better off setting this to empty (i.e. whole line commented out) and using the `default` function within the templates instead to set it to 'default' if not set?

Reply:
This is our best practice in the charts repo. We require new charts to do it like this. I think it makes sense. It's always better to have a default in `values.yaml`, if just for the sake of documentation. The `default` function should only be used if its not possible to add defaults to `values.yaml`.

Reply:
@unguiculus isn't it better to create a new ServiceAccount for `{{ template "cert-manager.fullname" . }}` by default rather than apply extra permissions to the `default` ServiceAccount. The is potentially a permission leak risk to other pods in the same namespace. Hence this should be blank as @munnerz suggests and default for the 'fullname' in the template.

Reply:
@whereisaaron This is actually what happens. If `rbac.create=true` a new servive account `{{ template "cert-manager.fullname" . }}` is created that gets all the permissions. If you don't use RBAC or want to manage RBAC resources outside the chart, set `rbac.create=false`. Then by default, the default service acount is used which is fine if you don't use RBAC. You may specify another service account with `rbac.serviceAccountName`.  As mentioned in the readme, `rbac.serviceAccountName` is ignored if `rbac.create=false`.

Reply:
Thanks for the explanation @unguiculus, according to the [Helm best practices documentation](https://docs.helm.sh/chart_best_practices/#using-rbac-resources) `rbac.serviceAccountName` must not be ignored. Chart authors are supposed to also support the situation where the user has externally managed RBAC resources. For example, let's say in my cluster I have created a Service Account 'foo' associated with the required RBAC roles for `cert-manager`. I should be able to deploy the chart with:
```
rbac.create=false
rbac.serviceAccountName="foo"
```

In this case the chart should not create RBAC resources (Roles etc.), but the Pod resources must be associated with the "foo" Service Account.

https://docs.helm.sh/chart_best_practices/#using-rbac-resources
> `rbac.serviceAccountName` should set the name of the ServiceAccount to be used by access-controlled resources created by the chart. If `rbac.create` is true, then a ServiceAccount with this name should be created. If `rbac.create` is false, then it should not be created, **but it should still be associated with the same resources** so that manually-created RBAC resources created later that reference it will function correctly. (Note that this effectively makes rbac.serviceAccountName a required value in these charts.)

Reply:
Thanks for pointing that out. So, the docs should be improved.

The initial PR that led to this best practice is here: https://github.com/coreos/alb-ingress-controller/pull/146. It was used as a reference for many PRs in the charts repo and was eventually put down as a best practice. `rbac.serviceAccountName` is and should be ignored if `rbac.create` is `true`. I'll try and create a PR to fix this.

Reply:
Thanks for the reference, the reasoning and practice seemed well thought out to me, so I'll be disappointed if it changes. The change would seem retrograde on two fronts:
1. It limits use cases, since externally managed permissions are no longer supported
2. I don't think it is intuitive. That's demonstrated by the need to have a comment to explain it's going to be ignored 😄  Without the comment, I didn't expect that.

Can you point me to the discussion that decided against the current best practice position? I'd like to understand their reasoning. I'm perplexed! 😕

In the end I highly appreciate consistency. So let's be consistent with whatever the core Helm team decide @munnerz

Reply:
I'm a little confused now. The way I implemented it here follows the current best practice. The docs are not correct. The best practice was brought up by @mgoodness in the mention PR and discussed in one of our calls. If I understand you correctly, the way I implemented it is exactly what you want. Your above-mentioned example works as you expect it to. It associates the service account `foo` and does not create RBAC resources, thus allowing you external managment of RBAC resources. See this line: https://github.com/jetstack/cert-manager/pull/229/files#diff-1f21f6872f614d29fa80979318b965d1R20

Reply:
Reading through this, I think we do need to change the RBAC best-practices docs, but I don't think serviceAccountName should ever be ignored if supplied.  Many organizations have rules around credential naming and a chart that ignores the supplied value and autogenerates a credential name can't support that.  serviceAccountName should be optional (rather than required as the docs state) with a default generated to be unique to the release (i.e. `foo.fullname`.

Reply:
@unguiculus I apologize, I misunderstood the case in which you want to ignore `serviceAccountName`. You are correct the PR preserves the ability to support externally managed RBAC/ServiceAccounts. I'm sorry I suggested otherwise above.

If I understand correctly now, you actually want to ignore the user-specified `serviceAccountName` when the chart is creating the RBAC resources?

I concur with @omkensey that a user-specified `serviceAccountName` should be used if supplied and not ignored.

I agree with @omkensey and @munnerz that if the `rbac.create=true` and `serviceAccountName` is not supplied or blank, that the chart should auto-generate a name (typically the `fullname` template). And I think the Best Practices docs should be updated to cover that situation.

The four possible cases for the current settings are as follows. `serviceAccountName` is optional but never ignored.

1. `rbac.create=true`, `rbac.serviceAccountname="foo"` ==> ServiceAccount "foo" created, pods use that ServiceAccount, RBAC resources created
2. `rbac.create=true`, `rbac.serviceAccountname=""` ==> ServiceAccount "{{ template 'fullname' . }}" created, pods use that ServiceAccount, RBAC resources created
3. `rbac.create=false`, `rbac.serviceAccountname="foo"` ==> No ServiceAccount created, but pods use ServiceAccount "foo", no RBAC resources created
4. `rbac.create=false`, `rbac.serviceAccountname=""` ==> No ServiceAccount created, pods use ServiceAccount "default", no RBAC resources created

The situation that doesn't cover, is if the user would like RBAC resources created but associated with an also existing Service account. To support that you'd need more options. Arguably ServiceAccounts are not part of RBAC anyway and should be separate consideration. (ServiceAccounts exist independently of RBAC, RBAC just piggy-backs on them. If you disable RBAC in a cluster you still have, need, and use ServiceAccounts) In that case you'd argue the Best Practice should really be separate settings. This is both conceptually cleaner and more accurate, and covers every use case any of us has discussed.

```
rbac:
  create: true

serviceAccount:
  create: false
  serviceAccountName: "foo"
```

Reply:
Yep I think I agree with @whereisaaron on this one - rbac and serviceAccounts are separate things and it does confuse things when they are conflated. Right now, if `rbac.create=true` and `rbac.serviceAccountName=foo`, I would expect no RBAC resources to be created **and** a service account named foo to **not** be created and instead used as the `serviceAccountName` on pods. Namely because I see `create` as a sort of toggle for the entire block (akin to `enabled`).

Splitting these two concepts out in the YAML does make this far clearer and disambiguates the two, and with sane defaults the UX is preserved (i.e. `rbac.create=true/false` behaves in a similar fashion to the current best practices).

Reply:
That said, I am keen to follow whatever is decided as best practice. This discussion is basically the blocker for merge now.

From what I can see, if this is accepted #232 can be closed?

Reply:
I think we should avoid automatically associating chart-created resources with the `default` service account as a best practice (basically defaulting to more secure rather than less).  It makes it hard to grant permissions to things without granting them to a whole hodgepodge of other things that don't need them.  If people want the default service account to be used, they should specify it explicitly.

I guess the remaining question is, is it worth adding the complexity of a new top-level key for service accounts to cover the case of "create RBAC resources that use an existing service account"?  Pro: My inner control-freak sysadmin says "yes, that's exactly what you'd do", and it allows creating and using multiple service accounts in a chart.  Con: it introduces new failure surface for syntax errors in the values file.

I'll start a discussion over in the Helm project and reference this thread, since this is really a "Helm chart best practices" discussion.

Reply:
@omkensey Let's discuss this in the next charts call.

Reply:
I think I'd add a layer of key-value abstraction to support multiple service accounts, e.g.:

```
serviceAccounts:
  client:
    serviceAccountName: client-account
    create: false
  server:
    serviceAccountName: server-account
    create: true
```